lib/block_transpose.in.cu

```diff
@@ -104,82 +115,70 @@ namespace quda
+  void launch_span_nColor(v_t &V, cvector_ref<b_t> &B, bool from_to_non_rel, IntList<nColor, N...>)
   {
     if (B[0].Ncolor() == nColor) {
-      block_transpose<v_t, b_t, vFloat, bFloat, nSpin, nColor>(V, B);
+      IntList<@QUDA_MULTIGRID_MRHS_LIST@> nVecs;
+      launch_span_nVec<v_t, b_t, vFloat, bFloat, nSpin, nColor>(V, B, from_to_non_rel, nVecs);
     } else {
-      IntList<N...> nVecs;
+      IntList<N...> nColors_remaining;
       if constexpr (sizeof...(N) > 0) {
-        launch_span_nColor<v_t, b_t, vFloat, bFloat, nSpin>(V, B, nVecs);
+        launch_span_nColor<v_t, b_t, vFloat, bFloat, nSpin>(V, B, from_to_non_rel, nColors_remaining);
       } else {
-        errorQuda("nColor = %d not instantiated", V.Ncolor());
+        errorQuda("nColor = %d not instantiated", B.Ncolor());
       }
     }
   }
 
-  template <class v_t, class b_t, typename vFloat, typename bFloat, int nSpin>
-  void block_transpose(v_t &V, cvector_ref<b_t> &B)
-  {
-    if (V.Ncolor() / V.Nvec() != B[0].Ncolor()) {
-      errorQuda("V.Ncolor() / V.Nvec() (=%d) != B.Ncolor() (=%d)", V.Ncolor() / V.Nvec(), B[0].Ncolor());
-    }
-
-    IntList<@QUDA_MULTIGRID_NVEC_LIST@> nColors;
-    launch_span_nColor<v_t, b_t, vFloat, bFloat, nSpin>(V, B, nColors);
-  }
-
-  template <class v_t, class b_t, typename vFloat, typename bFloat> void block_transpose(v_t &V, cvector_ref<b_t> &B)
+  template <class v_t, class b_t, typename vFloat, typename bFloat, int nSpin, int... N>
+  void launch_span_nSpin(v_t &V, cvector_ref<b_t> &B, bool from_to_non_rel, IntList<nSpin, N...>)
   {
     if (V.Nspin() != B[0].Nspin()) { errorQuda("V.Nspin() (=%d) != B.Nspin() (=%d)", V.Nspin(), B[0].Nspin()); }
 
-    if (V.Nspin() == 2) {
-      block_transpose<v_t, b_t, vFloat, bFloat, 2>(V, B);
-    } else if (V.Nspin() == 4) {
-      block_transpose<v_t, b_t, vFloat, bFloat, 4>(V, B);
-    } else if (V.Nspin() == 1) {
-      block_transpose<v_t, b_t, vFloat, bFloat, 1>(V, B);
+    if (V.Nspin() == nSpin) {
+      IntList<@QUDA_MULTIGRID_NC_NVEC_LIST@> nColors;
+      launch_span_nColor<v_t, b_t, vFloat, bFloat, nSpin>(V, B, from_to_non_rel, nColors);
     } else {
-      errorQuda("Unexpected nSpin = %d", V.Nspin());
+      if constexpr (sizeof...(N) > 0) {
+        IntList<N...> nSpins_remaining;
+        launch_span_nSpin<v_t, b_t, vFloat, bFloat>(V, B, from_to_non_rel, nSpins_remaining);
+      } else {
+        errorQuda("Unexpected nSpin = %d", V.Nspin());
+      }
     }
   }
 
-  template <class v_t, class b_t> void block_transpose(v_t &V, cvector_ref<b_t> &B)
+  template <class v_t, class b_t> void block_transpose(v_t &V, cvector_ref<b_t> &B, bool from_to_non_rel)
   {
     if (!is_enabled(V.Precision()) || !is_enabled(B[0].Precision()))
       errorQuda("QUDA_PRECISION=%d does not enable required precision combination (V = %d B = %d)", QUDA_PRECISION,
                 V.Precision(), B[0].Precision());
 
+    IntList<1, 2, 4> nSpins;
```

Comment:
I know there isn't a huge file to begin with, but this could be thinned down depending on which Dirac types are compiled in.

Reply:
https://github.com/lattice/quda/pull/1497/commits/cd3bbad7179f780979e7b6ffe2a88f71a5f18046